src/modules/roc_pipeline/receiver.cpp

```diff
@@ -35,8 +35,11 @@ Receiver::Receiver(const ReceiverConfig& config,
     , timestamp_(0)
     , num_channels_(packet::num_channels(config.common.output_channels))
     , active_cond_(control_mutex_) {
-    mixer_.reset(new (allocator_)
-                     audio::Mixer(sample_buffer_pool, config.common.internal_frame_size),
+    mixer_.reset(new (allocator_) audio::Mixer(
+                     sample_buffer_pool,
+                     packet::timestamp_from_ns(config.common.internal_frame_size,
+                                               config.common.output_sample_rate)
+                         * packet::num_channels(config.common.output_channels)),
```

Comment:
We're trying to keep roc_pipeline code very high-level. Usually it only constructs the pipeline but doesn't perform any calculations, low-level manipulations with parameters, etc.

So, as I mentioned in the issue, could you move the calculations inside the components constructors? Here and in all other places. So that, in particular, you'll pass config.common.internal_frame_size to Mixer and it will call timestamp_from_ns and num_channels internally.

Also, we should perform validation of the result. If the calculations result is zero, the corresponding component should log an error and switch to an an fail to construct. See, for example, audio::Watchdog.

This is how we deal with validation:

```
Mixer::Mixer(..., size_t internal_frame_size, ...)
    : valid_(false)
    , frame_size_(timestamp_from_ns(...) * num_channels(...))
{
    if (frame_size_ == 0) {
        roc_log(LogError, ...);
        return;
    }

    ...

    valid_ = true;
}
```

Then pipeline::Receiver will call mixer.valid() and see that mixer was failed to construct.

Reply:
So I'm thinking of modifying `Mixer` constructor by making it like 
```suggestion
Mixer(core::BufferPool<sample_t>& pool, 
          core::nanoseconds_t frame_size, 
          size_t sample_rate, 
          packet::channel_mask_t number_of_channels); 
```
But won't this also mean that I have to modify all calls made to this constructor `Mixer`, for example, in `src/modules/roc_audio/ text_mixer.cpp`, which just calls 
`Mixer mixer(buffer_pool, MaxSz);`
But this can be solved by making` size_t sample_rate, packet::channel_mask_t number_of_channels` optional by assigning them all to zero, making them default arguments.
So the final constructor would look like
```suggestion
Mixer(core::BufferPool<sample_t>& pool, 
         core::nanoseconds_t frame_size, 
         size_t sample_rate = 0, 
         packet::channel_mask_t number_of_channels = 0);
```
Can I do it like this?

Reply:
> packet::channel_mask_t number_of_channels

Note: it should be a channel mask, not number of channels. Then you can convert it to the number of channels using packet::num_channels().

> But won't this also mean that I have to modify all calls made to this constructor Mixer

Yep.

> But this can be solved by making size_t sample_rate, packet::channel_mask_t number_of_channels optional by assigning them all to zero, making them default arguments.

Formerly we used this approach, but then rejected it. The problem is that it was quite error-prone: it was easy to forget to pass optional arguments and create an inconsistency between pipeline elements (e.g. create two elements with different channel masks). So now we don't use default arguments for constructors and always explicitly provide all arguments.

Reply:
done.